lib/src/ignore.dart

```diff
@@ -0,0 +1,321 @@
+    end--;
+  }
+  // Empty patterns match nothing.
+  if (first == end) return null;
+
+  var current = first;
+  String peekChar() => current >= end ? null : pattern[current];
+
+  var expr = '';
+
+  // Parses the inside of a [] range. Returns the value as a RegExp character
+  // range, or null if the pattern was broken.
+  String parseCharacterRange() {
+    var characterRange = '';
+    var first = true;
+    for (;;) {
+      final nextChar = peekChar();
+      if (nextChar == null) {
+        print('a');
+        return null;
+      }
+      current++;
+      if (nextChar == '\\') {
+        final escaped = peekChar();
+        if (escaped == null) {
+          print('b');
+          return null;
+        }
+        current++;
+
+        characterRange += escaped == '-' ? r'\-' : RegExp.escape(escaped);
+      } else if (nextChar == '!' && first) {
+        characterRange += '^';
+      } else if (nextChar == ']' && first) {
+        characterRange += RegExp.escape(nextChar);
+      } else if (nextChar == ']') {
+        assert(!first);
+        return characterRange;
+      } else {
+        characterRange += nextChar;
+      }
+      first = false;
+    }
+  }
+
+  var relativeToPath = false;
+  var matchesDirectoriesOnly = false;
+
+  // slashes have different significance depending on where they are in
+  // the String. Handle that here.
+  // Returns the regexp to match the just parsed slash.
+  void handleSlash() {
+    if (current == end) {
+      // A slash at the end makes us only match directories.
+      matchesDirectoriesOnly = true;
+    } else {
+      // A slash anywhere else makes the pattern relative anchored at the
+      // current path.
+      relativeToPath = true;
+    }
+  }
+
+  for (;;) {
+    final nextChar = peekChar();
+    if (nextChar == null) break;
+    current++;
+    if (nextChar == '*') {
+      if (peekChar() == '*') {
+        // Handle '**'
+        current++;
+        if (peekChar() == '/') {
+          current++;
+          if (current == end) {
+            expr += '.*/';
+          } else {
+            // Match nothing or a path followed by '/'
+            expr += '(?:(?:)|(?:.*/))';
+          }
+          // Handle the side effects of seeing a slash.
+          // We know it was not initial, so the return value is ignored.
+          handleSlash();
+        } else {
+          expr += '.*';
+        }
+      } else {
+        // Handle a single '*'
+        expr += '[^/]*';
+      }
+    } else if (nextChar == '?') {
+      // Handle '?'
+      expr += '[^/]';
+    } else if (nextChar == '[') {
+      // Character ranges
+      final characterRange = parseCharacterRange();
+      if (characterRange == null) {
+        print('c');
+        return null;
+      }
+      expr += '[$characterRange]';
+    } else if (nextChar == '\\') {
+      // Escapes
+      final escaped = peekChar();
+      if (escaped == null) {
+        print('d');
+        return null;
```

Comment:
I have contemplated if we should throw a FormatException and then later up just discard those... But then we could have a pedantic parse function we can use to validate the patterns... Which might be nice in some cases.

Reply:
Yeah - Thought about the same.

I'll do that.

Reply:
I opted for returning a ParseResult that tells whether the pattern was parsed. And then taking a handler.